README.md
# Emote Racer 2 ![Website](https://img.shields.io/website?url=https%3A%2F%2Femoteracer.xyz)
## A fork of the original EmoteRacer
[![DigitalOcean Referral Badge](https://web-platforms.sfo2.digitaloceanspaces.com/WWW/Badge%203.svg)](https://www.digitalocean.com/?refcode=3470e3b80f7c&utm_campaign=Referral_Invite&utm_medium=Referral_Program&utm_source=badge)

![Game Screenshot](https://i.ibb.co/VJDpQfj/emoteracer-screenshot.jpg)

## Description
[Emote Racer 2](https://er.mmatt.net) is a free online multiplayer [Twitch Emote](https://www.twitch.tv/creatorcamp/en/learn-the-basics/emotes/) typing/guessing game, where you try to guess the name of a given emote. You can choose to play battle royale, 1 vs 1, or practice alone.

This was forked because the original Emote Racer was taken offline (and I assume not developed anymore). I am planning on adding more emotes and keeping it online :)

## Technologies
* JavaScript
* SCSS
* Node.js
* Express
* Socket.io

### `🙌` Code Contributors

<table>
<tr>
    <td align="center" style="word-wrap: break-word; width: 150.0; height: 150.0">
        <a href=https://github.com/mmattbtw>
            <img src=https://avatars.githubusercontent.com/u/30363562?v=4 width="100;"  style="border-radius:50%;align-items:center;justify-content:center;overflow:hidden;padding-top:10px" alt=matt/>
            <br />
            <sub style="font-size:14px"><b>matt</b></sub>
        </a>
    </td>
    <td align="center" style="word-wrap: break-word; width: 150.0; height: 150.0">
        <a href=https://github.com/carlssonk>
            <img src=https://avatars.githubusercontent.com/u/66022459?v=4 width="100;"  style="border-radius:50%;align-items:center;justify-content:center;overflow:hidden;padding-top:10px" alt=Oliver Carlsson/>
            <br />
            <sub style="font-size:14px"><b>Oliver Carlsson</b></sub>
        </a>
    </td>
    <td align="center" style="word-wrap: break-word; width: 150.0; height: 150.0">
        <a href=https://github.com/MrAuro>
            <img src=https://avatars.githubusercontent.com/u/35087590?v=4 width="100;"  style="border-radius:50%;align-items:center;justify-content:center;overflow:hidden;padding-top:10px" alt=Auro/>
            <br />
            <sub style="font-size:14px"><b>Auro</b></sub>
        </a>
    </td>
    <td align="center" style="word-wrap: break-word; width: 150.0; height: 150.0">
        <a href=https://github.com/Cryptizism>
            <img src=https://avatars.githubusercontent.com/u/60571306?v=4 width="100;"  style="border-radius:50%;align-items:center;justify-content:center;overflow:hidden;padding-top:10px" alt=Cryptizism/>
            <br />
            <sub style="font-size:14px"><b>Cryptizism</b></sub>
        </a>
    </td>
</tr>
</table>


CHANGELOG.md
# CHANGELOG

## [1.0.15](https://github.com/mmattDonk/emoteracer.xyz/releases/tag/1.0.15) - 2022-05-11 00:13:48

Clueless

## [v1.0.14](https://github.com/mmattDonk/emoteracer.xyz/releases/tag/v1.0.14) - 2022-02-19 05:57:33

## [v1.0.13](https://github.com/mmattDonk/emoteracer.xyz/releases/tag/v1.0.13) - 2022-02-17 22:12:48

*No description*

## [v1.0.12](https://github.com/mmattDonk/emoteracer.xyz/releases/tag/v1.0.12) - 2022-01-08 00:03:22

## [v1.0.11](https://github.com/mmattDonk/emoteracer.xyz/releases/tag/v1.0.11) - 2022-01-07 06:10:30

## [v1.0.10](https://github.com/mmattDonk/emoteracer.xyz/releases/tag/v1.0.10) - 2021-12-09 00:27:21

## [v1.0.9](https://github.com/mmattDonk/emoteracer.xyz/releases/tag/v1.0.9) - 2021-11-09 17:08:43

**Full Changelog**: https://github.com/mmattDonk/emoteracer.xyz/compare/v1.0.8...v1.0.9

## [v1.0.8](https://github.com/mmattDonk/emoteracer.xyz/releases/tag/v1.0.8) - 2021-11-09 17:06:53

## [v1.0.7](https://github.com/mmattDonk/emoteracer.xyz/releases/tag/v1.0.7) - 2021-10-05 16:00:49

**Full Changelog**: https://github.com/mmattDonk/emoteracer.xyz/compare/v1.0.6...v1.0.7

## [v1.0.6](https://github.com/mmattDonk/emoteracer.xyz/releases/tag/v1.0.6) - 2021-10-05 14:50:55

Updated deps.
## What's Changed
* chore(deps): bump axios from 0.21.3 to 0.21.4 by @dependabot in https://github.com/mmattDonk/emoteracer.xyz/pull/13
* Fetch Upstream by @mmattbtw in https://github.com/mmattDonk/emoteracer.xyz/pull/14
* chore(deps): bump axios from 0.21.4 to 0.22.0 by @dependabot in https://github.com/mmattDonk/emoteracer.xyz/pull/16


**Full Changelog**: https://github.com/mmattDonk/emoteracer.xyz/compare/v1.0.5...v1.0.6

## [v1.0.5](https://github.com/mmattDonk/emoteracer.xyz/releases/tag/v1.0.5) - 2021-09-05 05:30:32

Added a Discord button.

### Feature

- visual:
  - Added a Discord button ([986d80a](https://github.com/mmattDonk/emoteracer.xyz/commit/986d80aa2349660211eb33a4b956c73fcec727ea))

## [v1.0.4](https://github.com/mmattDonk/emoteracer.xyz/releases/tag/v1.0.4) - 2021-09-05 03:58:50

Added forsenDiscoSnake as a Cosmetic Profile Pic.

### Feature

- cosmetics:
  - Added forsenDiscoSnake as a Profile Picture ([d2f0aff](https://github.com/mmattDonk/emoteracer.xyz/commit/d2f0aff2a69ff2fefb2d27fe79e2b70ff406dc20))

## [v1.0.3](https://github.com/mmattDonk/emoteracer.xyz/releases/tag/v1.0.3) - 2021-09-04 19:31:25

Updated deps

## [v1.0.2](https://github.com/mmattDonk/emoteracer.xyz/releases/tag/v1.0.2) - 2021-09-04 18:48:52

Added LULE ([#10](https://github.com/mmattbtw/emoteracer.xyz/issues/10))

### Feature

- emotes:
  - Added LULE ([e22cd1f](https://github.com/mmattDonk/emoteracer.xyz/commit/e22cd1f79e722cc29b1fe0f139428ebccf4cecee))

### Documentation

- README:
  - Updated contributors ([799f760](https://github.com/mmattDonk/emoteracer.xyz/commit/799f76042c88a4a3a9c1598c72d8c03b6ac6aaf4))
  - Updated the support section. ([fed1748](https://github.com/mmattDonk/emoteracer.xyz/commit/fed1748f4cbaf44f6d2a553c9b8bf3c146b29321))
  - Added contributors ([b6963bb](https://github.com/mmattDonk/emoteracer.xyz/commit/b6963bb4200231445ec2239c9a7fb0fe88239574))

## [v1.0.1](https://github.com/mmattDonk/emoteracer.xyz/releases/tag/v1.0.1) - 2021-09-04 16:48:34

Added Enzo's emotes ([#5](https://github.com/mmattbtw/emoteracer.xyz/issues/5) and [#6](https://github.com/mmattbtw/emoteracer.xyz/pull/6))

### Feature

- emotes:
  - Added Enzo's Emote Suggestions ([9913c8d](https://github.com/mmattDonk/emoteracer.xyz/commit/9913c8d512c8e6f78fdba0e428e0399a5105739b)) ([#6](https://github.com/mmattDonk/emoteracer.xyz/pull/6))

## [v1.0.0](https://github.com/mmattDonk/emoteracer.xyz/releases/tag/v1.0.0) - 2021-09-04 16:25:55

*No description*

### Bug Fixes

- general:
  - fixed text ([1dcd3cd](https://github.com/mmattDonk/emoteracer.xyz/commit/1dcd3cdf88d4d49554f044d6c7f65f2835a29a97))
  - fixed player join leave bug ([48bdabd](https://github.com/mmattDonk/emoteracer.xyz/commit/48bdabdeb198170dd8fdf99ec05fd42574d9db00))
  - fixed lobby styling ([1c8f610](https://github.com/mmattDonk/emoteracer.xyz/commit/1c8f610f822cacad9e0f26f86a3d16b8ccc83259))

### Documentation

- prtemplate:
  - Add a Pull Request template. ([cc7aad5](https://github.com/mmattDonk/emoteracer.xyz/commit/cc7aad5fde50819b4f780d7c487f13b9347a0d06))

### Refactor

- general:
  - refactor ([2c6220b](https://github.com/mmattDonk/emoteracer.xyz/commit/2c6220b7b0528b0dcf35e8cba3b4d39ebf07c540))
  - refactoring & optimization ([07e9147](https://github.com/mmattDonk/emoteracer.xyz/commit/07e9147419aa2a5f40021dc49c750ac8154cf389))

\* *This CHANGELOG was automatically generated by [auto-generate-changelog](https://github.com/BobAnkh/auto-generate-changelog)*


emote_list_generator.js
const fs = require("fs");

let str = `...`;

const emoteArr = str.split("\n");
emoteArr.shift();

let final = [];

for (let emote of emoteArr) {
    let name = emote.split(": ")[0];
    let url = emote.split(": ")[1];

    if (url.match(/betterttv.com\/emotes\/[A-z0-9]+/)) {
        url = url.replace("betterttv.com/emotes/", "");
        final.push({
            id: url,
            name: name,
            provider: "bttv",
        });
    } else if (url.match(/frankerfacez.com\/emoticon\/[A-z0-9]+/)) {
        url = url.replace("frankerfacez.com/emoticon/", "");
        url = url.split("-")[0];
        final.push({
            id: url,
            name: name,
            provider: "ffz",
        });
    } else if (url.match(/7tv.app\/emotes\/[A-z0-9]+/)) {
        url = url.replace("7tv.app/emotes/", "");
        final.push({
            id: url,
            name: name,
            provider: "7tv",
        });
    }
}

fs.writeFileSync("./emotes.json", JSON.stringify(final));


public/js/OriginalGame.js
// #################################################################################
// ######################## GUESS THE EMOTE GAME (LOGIC) ###########################
// #################################################################################


function originalGame() {
  const emotes = getEmotes()

  let emotesClone = JSON.parse(JSON.stringify(emotes))
  let currentScore = 0;
  let currentEmote = {};
  let lives = 3
  let waitSubmit = true;

  const initRound = () => {
    emoteImg.src = "";
    gameUpperContent.style.visibility = "hidden";
    inputEmote.readOnly = false;
    inputEmote.focus(); // Automatically focus input text
    
    if(emotesClone.length <= 0) emotesClone = JSON.parse(JSON.stringify(emotes))
    lives = 3;

    const randomEmoteIndex = Math.floor(Math.random() * emotesClone.length)
    currentEmote = emotesClone[randomEmoteIndex]
    if(currentEmote.provider === "twitch") emoteImg.src = `https://static-cdn.jtvnw.net/emoticons/v1/${emotesClone[randomEmoteIndex].id}/3.0`
    if(currentEmote.provider === "bttv") emoteImg.src = `https://cdn.betterttv.net/emote/${emotesClone[randomEmoteIndex].id}/3x`
    if(currentEmote.provider === "ffz") emoteImg.src = `https://cdn.frankerfacez.com/emoticon/${emotesClone[randomEmoteIndex].id}/4`
    if(currentEmote.provider === "7tv") emoteImg.src = `https://cdn.7tv.app/emote/${emotesClone[randomEmoteIndex].id}/3x`


    emotesClone.splice(randomEmoteIndex, 1)
    $('.game-upper-content').waitForImages(function() {
      setTimeout(initRoundDOM, 400)
    });
  }

  const initRoundDOM = () => {
    gameUpperContent.style.visibility = "visible";
    gameUpperContent.classList.add("back-in-up-animation")
    setTimeout(() => waitSubmit = false, 400)
  }

  function guessListener(e) {
    if(e.key === "Enter") {
      if(inputEmote.value === "") return;
      while(waitSubmit === true) {
        waitSubmitDom();
        return;
      } 

      const guess = inputEmote.value.toLowerCase();
      const emoteCode = currentEmote.name.toLowerCase();
      
      if(guess === emoteCode) {
        handleRoundWin();
        inputEmote = document.querySelector(".inputEmote")
      } else {
        lives--
        wrongGuessAnimation()
        if(lives <= 0) {
          handleRoundLose();
        } 
      }
      inputEmote.value = "";
    }
  }

  function waitSubmitDom() {
    waitLabel.classList.remove("wait-label-fade")
    setTimeout(() => waitLabel.classList.add("wait-label-fade"), 50)
    setTimeout(() => waitLabel.classList.remove("wait-label-fade"), 850)
  }

  // DOM MANIPULATION WHEN USER GUESS WRONG
  const wrongGuessAnimation = () => {
    gameHeart[lives].classList.remove("fas"), gameHeart[lives].classList.add("far")
    inputEmote.classList.add("input-wrong");
    inputEmote.classList.add("animate__headShake");

    // setTimeout(() => inputEmote.classList.remove("animate__headShake"), 400)
    setTimeout(function() {
      inputEmote.classList.remove("animate__headShake")
      inputEmote.classList.remove("input-wrong");
    }, 400)
  }

  // HANDLE CORRECT GUESS
  const handleRoundWin = () => {
    waitSubmit = true;
    currentScore++;
    emoteName.innerText = currentEmote.name;
    emoteName.classList.add("show-fade")

    for(let score of finalScore) score.innerText = currentScore;

    setTimeout(function() {
      nextEmoteUpperAnimation();
      setTimeout(initRound, 400);
    }, 1000)
  }


  // HANDLE ROUND LOSE
  const handleRoundLose = () => {
    waitSubmit = true;
    inputEmote.readOnly = true;
    inputEmote.value = "";
    emoteName.innerText = currentEmote.name;
    emoteName.classList.add("show-fade")
    gameResults.classList.add("animate__fadeIn")
    gameResults.classList.remove("hide")
  }

  // Play again button execute
  const playAgainDomReset = () => {
    inputEmote.focus(); // Automatically focus input text

    gameResults.classList.remove("animate__fadeIn") // Resets game results popup menu
    for(let score of finalScore) score.innerText = currentScore; // Reset score

    nextEmoteUpperAnimation();
    setTimeout(initRound, 400)
  }

  const nextEmoteUpperAnimation = () => {
    for(let heart of gameHeart) heart.classList.add("fas"), heart.classList.remove("far")
    
    gameUpperContent.classList.add("animate__backOutRight")
    setTimeout(function() {
      gameUpperContent.classList.remove("animate__backOutRight")
      // gameUpperContent.classList.add("animate__backInLeft")

      gameUpperContent.classList.remove("back-in-up-animation");

      emoteName.classList.remove("show-fade");
      emoteName.innerText = "";
    }, 800)
  }


  // #################################################################################
  // ############### GUESS THE EMOTE GAME (DYNAMIC PAGE CREATION) ####################
  // #################################################################################


  this.originalGameHTML = function() {
    return (
      game.innerHTML =
      `
      <div class="original-game">
      <div class="game-results animate__animated animate__faster hide">
        <div class="new-personal-best hide">New Personal Best!</div>
        <h2>Your Score:</h2>
        <div class="final-score final-score-result">0</div>
        <div class="share-container">
          <a class="share-btn twitter-share-btn" href="https://twitter.com/share" target="_blank">Tweet</a>
          <a class="share-btn facebook-share-btn" href="https://www.facebook.com/sharer/sharer.php?u=#url" target="_blank">Share</a>
        </div>
        <div class="game-results-nav">
          <button class="game-nav-btn play-again-btn">PLAY AGAIN</button>
          <button class="game-nav-btn main-lobby-btn">MAIN MENU</button>
        </div>
      </div>
        <div class="game-upper">
          <div class="heart-container">
            <i class="fas fa-heart game-heart"></i>
            <i class="fas fa-heart game-heart"></i>
            <i class="fas fa-heart game-heart"></i>
          </div> 
          <h2 class="final-score-game">SCORE: <span class="final-score">0</span></h2>
           <div class="game-upper-content-container">
            <div class="game-upper-content animate__animated animate__fast">
              <div class="emote-img-container">
                <img class="emote-img" src="" alt="">
                <h1 class="emote-name"></h1>
              </div>
            </div>
          </div>
        </div>
      <div class="game-lower">
        <div class="emote-container">
          <div class="wait-label">WAIT FOR NEXT EMOTE</div> 
          <input
          class="inputEmote animate__animated animate__faster"
          type="text"
          maxlength="30"
          placeholder="TYPE HERE"
          onfocus="this.placeholder = ''"
          onblur="this.placeholder = 'TYPE HERE'"
          autofocus>
        </div>
        <h2 class="game-lower-guess-the-name">GUESS THE NAME OF ABOVE EMOTE</h2>
      </div>
    </div>
      `
    )
  }

  let emoteImg = null;
  let gameResults = null;
  let emoteName = null;
  let gameUpperContent = null;
  let finalScore = null;
  let playAgainBtn = null;
  let mainLobbyBtn = null;
  let inputEmote = null;
  let gameHeart = null;
  let waitLabel = null;
  this.originalGameDOM = function() {
  gameHeart = document.querySelectorAll(".game-heart");
  inputEmote = document.querySelector(".inputEmote");
  emoteImg = document.querySelector(".emote-img");
  emoteName = document.querySelector(".emote-name");
  gameResults = document.querySelector(".game-results");
  gameUpperContent = document.querySelector(".game-upper-content");
  finalScore = document.querySelectorAll(".final-score");
  playAgainBtn = document.querySelector(".play-again-btn");
  mainLobbyBtn = document.querySelector(".main-lobby-btn");
  waitLabel = document.querySelector(".wait-label");
  }


  this.originalGameEVENT = function() {
    inputEmote.addEventListener("keypress", guessListener)

    // Play Again
    playAgainBtn.addEventListener("click", function() {
      currentScore = 0;
      gameResults.classList.add("hide")
      playAgainDomReset();
    });

    // Back To Main Page
    mainLobbyBtn.addEventListener("click", function() {
      mainPage();
    });
  }

  pageChangeDisplay("game")

  originalGameHTML(); // Loads html
  pageTransition(); // Page transition
  originalGameDOM(); // Inits dom wiring
  originalGameEVENT(); // Inits event listeners

  inputEmote.focus(); // Automatically focus input text
  gameUpperContent.style.visibility = "hidden";
  // setTimeout(initRound, 1000) // Start game script
  initRound()
}

public/js/RacerGame.js
// #################################################################################
// ######################## GUESS THE EMOTE GAME (LOGIC) ###########################
// #################################################################################


function racerGame() {
  const emotes = getEmotes()

  let currentPage = "racer-game";

  // --Stats--
  // Speed
  let speed = 0;
  // Accuracy
  const alphabet = ["a","b","c","d","e","f","g","h","i","j","k","l","m","n","o","p","q","r","s","t","u","v","w","x","y","z", "0", "1", "2", "3", "4", "5", "6", "7", "8", "9"]
  let roundKeysTyped = 0;
  let totalKeysTyped = 0;
  let totalEmoteCharacters = 0;
  // Incorrect Guesses
  let incorrectGuesses = 0;
  // Game
  let emotesClone = JSON.parse(JSON.stringify(emotes))
  let currentScore = 0;
  let currentEmote = {};
  let gameStarted = false;
  // So we can PRELOAD Second Image
  let cycleBool = false;
  let currentEmoteFirst = {};
  let currentEmoteLast = {};
  let randomEmoteIndex = 0;
  let randomEmoteIndex2 = 0;

  function timer() {
    // Configutation for game
    startGameConfig()

    // Show timer
    racerGameTimerContainer.classList.remove("hide-fade");
    racerGameTimerContainer.style.display = ""
    
    // Timer logic
    let count = 60;
    let counter = setInterval(() => initBattleRoyaleTimer(), 1000); //10 will  run it every 100th of a second
    function initBattleRoyaleTimer() {
      count--;
      racerGameTimer.innerHTML=count + "s"; 
      if(count <= 0) {
        // When count hits 1, exec handleRoundEnd
        handleRoundEnd();
        clearInterval(counter)
        return;
      } 
    }
    initRound();
  }

  function startGameConfig() {
    gameStarted = true;
    gameInfoLabelContainer.style.display = "none"
    gameUpperContent.style.transition = "100ms"
    circleBox.classList.add("racer-timer-bar")

    // Generate For First slot
    randomEmoteIndex = Math.floor(Math.random() * emotesClone.length)
    currentEmoteFirst = emotesClone[randomEmoteIndex]
    emotesClone.splice(randomEmoteIndex, 1)
  }

  const initRound = () => {
    cycleBool = !cycleBool
    roundKeysTyped = 0;
    inputEmote.readOnly = false;
    inputEmote.focus(); // Automatically focus input text

    gameUpperContent.style.marginTop = "500px"
    setTimeout(() => gameUpperContent.style.transform = "", 50)
    setTimeout(() => gameUpperContent.style.clip = "", gameUpperContent.style.position = "", 100)


    if(emotesClone.length <= 0) emotesClone = JSON.parse(JSON.stringify(emotes))

    // SHOW FIRST SLOT
    slotVisibility(cycleBool);

    // Set opposite img depending on bool
    setImg(cycleBool);

    // GENERATE NEW FOR OTHER SLOT
    generateImg(cycleBool);


    setTimeout(() => initRoundDOM(), 100)
  }


  function slotVisibility(cycleBool) {
    let num1;
    let num2;
    if(cycleBool === true) num1 = 0, num2 = 1;
    if(cycleBool === false) num1 = 1, num2 = 0;

    emoteImg[num1].style.height = "";
    emoteImg[num1].style.width = "";
    emoteImg[num1].style.visibility = "visible";
    emoteImg[num2].style.visibility = "hidden";
    emoteImg[num2].style.height = "0";
    emoteImg[num2].style.width = "0";
  }

  function setImg(cycleBool) {
    if(cycleBool === true) {
      setEmoteFirst()
      currentEmote = currentEmoteFirst;
    } else {
      setEmoteLast();
      currentEmote = currentEmoteLast;
    }
  }

  function generateImg(cycleBool) {
    if(cycleBool === true) {
      randomEmoteIndex2 = Math.floor(Math.random() * emotesClone.length)
      currentEmoteLast = emotesClone[randomEmoteIndex2]

      setEmoteLast();

      emotesClone.splice(randomEmoteIndex2, 1)
    } else {
      randomEmoteIndex = Math.floor(Math.random() * emotesClone.length)
      currentEmoteFirst = emotesClone[randomEmoteIndex]
  
      setEmoteFirst();

      emotesClone.splice(randomEmoteIndex, 1)
    }
  }

  function setEmoteFirst() {
    if(currentEmoteFirst.provider === "twitch") emoteImg[0].src = `https://static-cdn.jtvnw.net/emoticons/v1/${currentEmoteFirst.id}/3.0`
    if(currentEmoteFirst.provider === "bttv") emoteImg[0].src = `https://cdn.betterttv.net/emote/${currentEmoteFirst.id}/3x`
    if(currentEmoteFirst.provider === "ffz") emoteImg[0].src = `https://cdn.frankerfacez.com/emoticon/${currentEmoteFirst.id}/4`
    if(currentEmoteFirst.provider === "7tv") emoteImg[0].src = `https://cdn.7tv.app/emote/${currentEmoteFirst.id}/3x`

  }

  function setEmoteLast() {
    if(currentEmoteLast.provider === "twitch") emoteImg[1].src = `https://static-cdn.jtvnw.net/emoticons/v1/${currentEmoteLast.id}/3.0`
    if(currentEmoteLast.provider === "bttv") emoteImg[1].src = `https://cdn.betterttv.net/emote/${currentEmoteLast.id}/3x`
    if(currentEmoteLast.provider === "ffz") emoteImg[1].src = `https://cdn.frankerfacez.com/emoticon/${currentEmoteLast.id}/4`
    if(currentEmoteLast.provider === "7tv") emoteImg[1].src = `https://cdn.7tv.app/emote/${currentEmoteLast.id}/3x`
  }


  const initRoundDOM = () => {
    gameUpperContent.style.marginTop = "0px"

    // SPEED TIMER
    setTimeout(function() {
      const speedInterval = setInterval(() => currentSpeed(), 10)
      function currentSpeed() {
        speed += 10;
      }
      this.stopSpeed = function() {
        clearInterval(speedInterval)
      }
    }, 90)
  }

  function guessListener(e) {
    if(gameStarted === false && e.key === "Enter") timer(); // <--- inits game
    if(gameStarted === true && e.key === "Enter") {
      if(inputEmote.value === "") return;
      const guess = inputEmote.value.toLowerCase();
      const emoteCode = currentEmote.name.toLowerCase();
      
      if(guess === emoteCode) {
        // STATS
        totalKeysTyped += roundKeysTyped; // GET TOTAL KEYS TYPED
        totalEmoteCharacters += currentEmote.name.length; // GET TOTAL EMOTE CHARACTERS
        stopSpeed(); // STOP SPEED TIMER

        handleRoundWin();
        inputEmote = document.querySelector(".inputEmote")
      } else {
        incorrectGuesses++; // Incorrect Guesses
        wrongGuessAnimation()
      }
      inputEmote.value = "";
    }

    // ADD ROUND KEYS TYPED
    if(alphabet.includes(e.key.toLowerCase())) {
      roundKeysTyped++;
    }
  }

  let skipBool = true;
  function skipListener(e) {
    if(skipBool) {
      skipBool = false;
 
      if(gameStarted === true && e.key === "ArrowUp") {
        inputEmote.value = ""
        gameUpperContent.style.transform = "scale(0)"
  
        setTimeout(function() {
          gameUpperContent.style.clip = "rect(0px,0px,0px,0px)", gameUpperContent.style.position = "absolute";
          initRound("skip");
        }, 100)
      } 
      
      setTimeout(() => skipBool = true, 400)
    }
  }


  // DOM MANIPULATION WHEN USER GUESS WRONG
  const wrongGuessAnimation = () => {
    inputEmote.classList.add("input-wrong");
    inputEmote.classList.add("animate__headShake");

    setTimeout(function() {
      inputEmote.classList.remove("animate__headShake")
      inputEmote.classList.remove("input-wrong");
    }, 400)
  }

  // HANDLE CORRECT GUESS
  const handleRoundWin = () => {
    // Add score
    currentScore++;
    for(let score of finalScore) score.innerText = currentScore;

    // Emote animation
    gameUpperContent.style.transform = "scale(0)"

    setTimeout(function() {
      gameUpperContent.style.clip = "rect(0px,0px,0px,0px)", gameUpperContent.style.position = "absolute";
      initRound()
    }, 100)

  }


  // HANDLE ROUND END
  const handleRoundEnd = () => {
    gameStarted = false;
    inputEmote.readOnly = true;
    inputEmote.value = "";
    gameResults.style.display = "block";
    gameResults.classList.add("racer-fade-output")
    gameUpperContent.style.marginTop = "500px";

    // CALCULATE STATS
    const avgSpeed = speed / currentScore;
    const avgSpeedText = `${Math.round(avgSpeed) / 1000}s`
    resultSpeedStats.innerText = avgSpeedText;

    const accuracy = (totalEmoteCharacters / totalKeysTyped) * 100;
    const accuracyText = `${Math.round((accuracy + Number.EPSILON) * 100) / 100}%`
    resultAccuracyStats.innerText = accuracyText;

    resultIncorrectStats.innerText = incorrectGuesses;

    // Remove event listeners
    document.removeEventListener("keypress", guessListener)
    document.removeEventListener("keydown", skipListener)

  }

  // Play again button execute
  const playAgainDomReset = () => {
    inputEmote.focus(); // Automatically focus input text
    for(let score of finalScore) score.innerText = currentScore; // Reset score
    gameUpperContent.style.clip = "rect(0px,0px,0px,0px)", gameUpperContent.style.position = "absolute";

    gameInfoLabelContainer.style.display = "block"
    circleBox.classList.remove("racer-timer-bar")

    // Reset Stats
    speed = 0;
    roundKeysTyped = 0;
    totalKeysTyped = 0;
    totalEmoteCharacters = 0;
    incorrectGuesses = 0;

    // Add event listener
    document.addEventListener("keypress", guessListener)
    document.addEventListener("keydown", skipListener)
  }


  // #################################################################################
  // ############### GUESS THE EMOTE GAME (DYNAMIC PAGE CREATION) ####################
  // #################################################################################


  this.racerGameHTML = function() {
    return (
      game.innerHTML =
      `
      <div class="original-game">
      <div class="game-results animate__animated animate__faster" style="display: none">
        <div class="new-personal-best hide">New Personal Best!</div>
        <h2>Your Score:</h2>
        <div class="final-score final-score-result">0</div>
        <div class="result-stats">
          <h3>STATS</h3>
          <div>Speed: <span class="result-speed-stats"></span></div>
          <div>Accuracy: <span class="result-accuracy-stats"></span></div>
          <div>Incorrect Guesses: <span class="result-incorrect-stats"></span></div>
        </div>
        <div class="game-results-nav">
          <button class="game-nav-btn play-again-btn">PLAY AGAIN</button>
          <button class="game-nav-btn main-lobby-btn">MAIN MENU</button>
        </div>
      </div>
      <div class="game-info-label-container">
        <h2>GUESS AS MANY EMOTES AS YOU CAN UNDER 60 SECONDS</h2>
        <h2 class="racer-start-game-label"> PRESS ENTER <img class="enter-key" src="imgs/EnterKey.svg"> TO START</h2>
        <button class="racer-start-game-label-btn" style="display: none;">PRESS TO START</button>
      </div>
      <div class="game-upper">
        <div class="racer-game-timer-container">
          <div class="racer-game-timer">60s</div>
          <svg class="circle-box" width="100%" height="100%">
            <circle cx="50%" cy="50%" r="47" stroke-width="6" fill="transparent"/>
          </svg>
        </div>
        <h2 class="final-score-game final-score-racer">SCORE: <span class="final-score">0</span></h2>
          <div class="game-upper-content-container">
          <div class="game-upper-content animate__animated animate__fast" style="margin-top: 500px">
            <div class="emote-img-container">
              <img class="emote-img" src="" alt="">
              <img class="emote-img" src="" alt="">
            </div>
          </div>
        </div>
      </div>
      <div class="game-lower">
        <div class="emote-container">
          <div class="wait-label">WAIT FOR NEXT EMOTE</div> 
          <input
          class="inputEmote animate__animated animate__faster"
          type="text"
          maxlength="30"
          placeholder="TYPE HERE" 
          onfocus="this.placeholder = ''"
          onblur="this.placeholder = 'TYPE HERE'"
          autofocus
          readonly>
        </div>
        <div class="game-skip-container">
          <h2 class="game-skip-label">PRESS <img class="arrow-up-key"src="imgs/ArrowUp.svg"> (Up Arrow) TO SKIP</h2>
        </div>
      </div>
    </div>
      `
    )
  }
  let racerStartGameLabelBtn = null;
  let emoteImg = null;
  let gameResults = null;
  let emoteName = null;
  let gameUpperContent = null;
  let finalScore = null;
  let playAgainBtn = null;
  let mainLobbyBtn = null;
  let inputEmote = null;
  let waitLabel = null;
  let racerGameTimerContainer = null;
  let racerGameTimer = null;
  let gameInfoLabelContainer = null;
  let circleBox = null;
  // stats
  let resultSpeedStats = null;
  let resultAccuracyStats = null;
  let resultIncorrectStats = null;
  this.racerGameDOM = function() {
  racerStartGameLabelBtn = document.querySelector(".racer-start-game-label-btn");
  inputEmote = document.querySelector(".inputEmote");
  emoteImg = document.querySelectorAll(".emote-img");
  emoteName = document.querySelector(".emote-name");
  gameResults = document.querySelector(".game-results");
  gameUpperContent = document.querySelector(".game-upper-content");
  finalScore = document.querySelectorAll(".final-score");
  playAgainBtn = document.querySelector(".play-again-btn");
  mainLobbyBtn = document.querySelector(".main-lobby-btn");
  waitLabel = document.querySelector(".wait-label");
  racerGameTimerContainer = document.querySelector(".racer-game-timer-container");
  racerGameTimer = document.querySelector(".racer-game-timer");
  gameInfoLabelContainer = document.querySelector(".game-info-label-container")
  circleBox = document.querySelector(".circle-box")
  // stats
  resultSpeedStats = document.querySelector(".result-speed-stats")
  resultAccuracyStats = document.querySelector(".result-accuracy-stats")
  resultIncorrectStats = document.querySelector(".result-incorrect-stats")
  }


  this.racerGameEVENT = function() {
    // inputEmote.addEventListener("keypress", guessListener)
    document.addEventListener("keypress", guessListener)
    document.addEventListener("keydown", skipListener)

    // Play Again
    playAgainBtn.addEventListener("click", function() {
      racerGameRemoveEVENT();
      currentScore = 0;
      gameResults.style.display = "none"
      playAgainDomReset();
    });

    // Back To Main Page
    mainLobbyBtn.addEventListener("click", function() {
      racerGameRemoveEVENT();
      mainPage();
    });

    racerStartGameLabelBtn.addEventListener("click", function() {
      timer();
    })
    
  }

  this.racerGameRemoveEVENT = function() {
    // remove potential event listeners from last session
    if(typeof guessListener === "function") document.removeEventListener("keypress", guessListener)
    if(typeof skipListener === "function") document.removeEventListener("keydown", skipListener)
  }

  pageChangeDisplay("game")

  racerGameHTML(); // Loads html
  pageTransition("game"); // Page transition
  racerGameDOM(); // Inits dom wiring
  racerGameEVENT(); // Inits event listeners

  inputEmote.focus(); // Automatically focus input text
  gameUpperContent.style.clip = "rect(0px,0px,0px,0px)", gameUpperContent.style.position = "absolute";

  this.getCurrentPage = function() {
    return currentPage;
  }
}